src/vs/workbench/parts/tasks/electron-browser/terminalTaskSystem.ts

```diff
@@ -867,7 +869,20 @@ export class TerminalTaskSystem implements ITaskSystem {
 				delete this.sameTaskTerminals[taskKey];
 			}
 		} else if (allowsSharedTerminal) {
-			let terminalId = this.idleTaskTerminals.remove(taskKey) || this.idleTaskTerminals.shift();
+			// Always allow to reuse the terminal previously used by the same task.
+			let terminalId = this.idleTaskTerminals.remove(taskKey);
+			if (!terminalId) {
+				// There is no idle terminal which was used by the same task.
+				// Search for any idle terminal used previously by a task of the same terminalGroup
+				// (or, if the task has no terminalGroup, a terminal used by a task without terminalGroup).
+				for (const taskId of this.idleTaskTerminals.keys()) {
+					const idleTerminalId = this.idleTaskTerminals.get(taskId)!;
```

Comment:
This loop needs to access both key and value of the `idleTaskTerminals` LinkedMap. Maybe it is worth adding an `entries(): [V, K][]` method to the LinkedMap class to avoid usage of the `!` operator? WDYT?
https://github.com/Microsoft/vscode/blob/ac2c2922dc1ce64e694c0b6d024b68af42333020/src/vs/base/common/map.ts#L499

```ts


	entries(): [V, K][] {
		let result: [V, K][] = [];
		let current = this._head;
		while (current) {
			result.push([current.value, current.key]);
			current = current.next;
		}
		return result;
	}
```